lib/msf/base/sessions/command_shell.rb

```diff
@@ -276,19 +277,40 @@ def cmd_shell(*args)
 
     # Why `/bin/sh` not `/bin/bash`, some machine may not have `/bin/bash` installed, just in case.
     # 1. Using python
-    # 1.1 Check Python installed or not
-    # We do not need to care about the python version
-    # Beacuse python2 and python3 have the same payload of spawn a shell
     python_path = binary_exists("python")
     if python_path != nil
+      # Payload: import pty;pty.spawn('/bin/bash')
+      # Base64 encoded payload: aW1wb3J0IHB0eTtwdHkuc3Bhd24oJy9iaW4vYmFzaCcp
       # Payload: import pty;pty.spawn('/bin/sh')
       # Base64 encoded payload: aW1wb3J0IHB0eTtwdHkuc3Bhd24oJy9iaW4vc2gnKQ==
       print_status("Using `python` to pop up an interactive shell")
-      shell_command("#{python_path} -c 'exec(\"aW1wb3J0IHB0eTtwdHkuc3Bhd24oJy9iaW4vc2gnKQ==\".decode(\"base64\"))'")
+      bash_path = binary_exists("bash")
+      if bash_path != nil
+        shell_command("#{python_path} -c 'exec(\"aW1wb3J0IHB0eTtwdHkuc3Bhd24oJy9iaW4vYmFzaCcp\".decode(\"base64\"))'")
+      else
+        shell_command("#{python_path} -c 'exec(\"aW1wb3J0IHB0eTtwdHkuc3Bhd24oJy9iaW4vc2gnKQ==\".decode(\"base64\"))'")
+      end
+      return
+    end
+	
+    # 2. Using python3
+    python3_path = binary_exists("python3")
+    if python3_path != nil
+      # Payload: import pty;pty.spawn('/bin/bash')
+      # Base64 encoded payload: aW1wb3J0IHB0eTtwdHkuc3Bhd24oJy9iaW4vYmFzaCcp
+      # Payload: import pty;pty.spawn('/bin/sh')
+      # Base64 encoded payload: aW1wb3J0IHB0eTtwdHkuc3Bhd24oJy9iaW4vc2gnKQ==
+      print_status("Using `python3` to pop up an interactive shell")
+      bash_path = binary_exists("bash")
+      if bash_path != nil
+        shell_command("#{python3_path} -c 'import base64;exec(base64.b64decode(\"aW1wb3J0IHB0eTtwdHkuc3Bhd24oJy9iaW4vYmFzaCcp\"))'")
+      else
+        shell_command("#{python3_path} -c 'import base64;exec(base64.b64decode(\"aW1wb3J0IHB0eTtwdHkuc3Bhd24oJy9iaW4vc2gnKQ==\"))'")
+      end
       return
     end
 
-    # 2. Using script
+    # 3. Using script
```

Comment:
This should really unify the Python version support in the way that payloads do. Newer versions of Linux are shipping Python3 as Python which is the case on my test system. That leads to a bug where since the `python` binary exists, it's assumed to be Python 2 (which is not save) causing a string decode error to be raised.

![image](https://user-images.githubusercontent.com/58950994/91723578-f273f300-eb69-11ea-9c61-8c218fbc154f.png)

I would instead suggest an approach like the following which is version independent. The only catch is you'll need to expose the `create_exec_stub` as a class method by updating `lib/msf/core/payload/python.rb`.
```suggestion
      print_status("Using `python` to pop up an interactive shell")
      shell_path = binary_exists("bash") || '/bin/sh'
      shell_command("#{python_path} -c \"#{ Msf::Payload::Python.create_exec_stub("import pty; pty.spawn('#{shell_path}')") } \"")
      return
    end

    # 2. Using script
```

<details>
<summary>lib/msf/core/payload/python.rb changes</summary>

Exposes the `py_create_exec_stub` as the `Msf::Payload::Python.create_exec_stub` class method so it can be used elsewhere.

```ruby
# -*- coding: binary -*-
require 'msf/core'

module Msf::Payload::Python

  #
  # Encode the given python command in base64 and wrap it with a stub
  # that will decode and execute it on the fly. The code will be condensed to
  # one line and compatible with all Python versions supported by the Python
  # Meterpreter stage.
  #
  # @param cmd [String] The python code to execute.
  # @return [String] Full python stub to execute the command.
  #
  def self.create_exec_stub(cmd)
    # Base64 encoding is required in order to handle Python's formatting
    b64_stub = "exec(__import__('base64').b64decode(__import__('codecs').getencoder('utf-8')('#{Rex::Text.encode_base64(cmd)}')[0]))"
  end

  def py_create_exec_stub(cmd)
    Msf::Payload::Python.create_exec_stub(cmd)
  end

end
```
</details>

Reply:
Also to address the original issue, set `python_path` with `python_path = binary_exists("python") || binary_exists("python3")` to fail back to `python3` if `python` does not exist.

Reply:
> Newer versions of Linux are shipping Python3 as Python which is the case on my test system.

Aaah, I was not aware that this was happening. In my experience, all python3 binaries were labeled as python3 so I had not come across this scenario.

Whilst I would commit your changes to the patch directly, it seems that Github does not support applying suggestions on deleted lines, so I will need to manually make the changes to the files.

Reply:
There was a spare `b64_stub` line in `python.rb` that was added by the original author which seems unused to me, although I am not familiar enough with the specifics of the language itself to remove it, so I have left it as is.